#### 에러/예외 처리

* 디버깅
* 에러와 예외
* 예외 처리
* 예외 발생 시키기



가. 디버깅 : 잘못된 프로그램을 수정하는 것 ex) 조건/반복할 때 많이 발생(index 에러(빈 리스트에 인덱스 찾을 수 없음) 등) => print 함수 활용!!

나. 에러와 예외

1. 문법 에러(Syntax Error) : 발생하면 파이썬 실행 안 됨/ 파일 이름, 줄번호, ^문자(줄에서 에러가 감지된 가장 앞의 위치를 가리키는 캐럿 기호(^) 표시)를 통해 파이썬이 코드를 읽어 나갈 때 문제가 발생한 위치를 표현

   가) Invalid syntax(문법 오류)

   ```python
   while
   ```

   나) assign to literal(잘못된 할당)

   ```python
   5=3
   # sum = 5하면 이후 sum()함수 동작 안 함
   ```

   다) EOL(End of Line)

   ```python
   print('hello
   ```

   라) EOF(End of File)

   ```python
   print(
   ```

2. 예외(Exception)

   가) 실행 도중 예상치 못한 상황을 맞이하면, 프로그램 실행을 멈춤

     1)문장이나 표현식이 문법적으로 올바르더라도 발생하는 에러

   나) 실행 중에 감지되는 에러들을 예외라고 부름

   다) 예외는 여러 타입으로 나타나고, 타입이 메시지의 일부로 출력됨

     1)NameError, TypeError 등은 발생한 예외 타입의 종류(이름)

   라) 모든 내장 예외는 Exception Class를 상속받아 이뤄짐

   마) 사용자 정의 예외를 만들어 관리할 수 있음

   바) ZeroDivisionError : 0으로 나누고자 할 때 발생

   ```python
   10/0
   ```

   사) NameError : namespace 상에 이름이 없는 경우

   ```python
   print(name_error) # 'name_error' is not defined
   ```

   아) TypeError

   ```python
   # 타입 불일치
   1 + '1' # 'int' and 'str' # 숫자와 문자 더하기
   
   round('3.5') # type str doesn't define __round__ method # 문자열 반올림할 때
   
   # argument 누락
   divmod() # divmod expected 2 arguments, got 0 # 인자 누락 시
   
   import random
   random.sample() # sample() missing 2 required positional arguments
   
   # argument 개수 초과
   divmod(1, 2, 3) # divmod expected 2 arguments, got 3 # 인자 개수 초과해도 안 됨
   
   import random
   random.sample(range(3), 1, 2) # sample() takes 3 positional arguments, but 4 were given
   
   # argument type 불일치
   import random
   random.sample(1, 2) # Population(첫 인자) must be a sequence. # 첫 인자는 시퀀스 들고와야 함
   ```

   자) ValueError

   ```python
   # 타입은 올바르나 값이 적절하지 않거나 없는 경우
   int('3.5') 
   
   range(3).index(6) # 6 is not in range # 6이 없는데요?(find는 -1로 반환)
   ```

   차) IndexError

   ```python
   # 인덱스가 존재하지 않거나 범위를 벗어나는 경우
   empty_list = []
   empty_list[2] # list index out of range : 가봤는데 이런 방은 없던데요?
   ```

   카) KeyError

   ```python
   # 해당 키가 존재하지 않는 경우
   song = {'IU' : '좋은날'}
   song['BTS'] # 없는데요? # get이면 에러 안 남
   ```

   타) ModuleNotFoundError

   ```python
   import yuhyun # No module named 'yuhyun' # 모듈 없던데요?
   ```

   파) ImportError

   ```python
   # Module은 있으나 존재하지 않는 클래스/함수를 가져오는 경우
   from random import sample
   print(sample(range(3), 1)) # [1]
   
   from random import samp # cannot import name 'samp' from 'random'
   ```

   하) KeyboardInterrupt : 임의로 프로그램을 종료했을 때

   거) IndentationError

   ```python
   # Indentation이 적절하지 않는 경우(네 칸 안 띄우거나)
   for i in range(3):
   print(i) # 네칸 안띄어서 에러
   
   for i in range(3):
       print(i)
       	print(i) # unexpected indent
   ```

다. 예외 처리

1. try 문(statement) / except 절(clause)을 이용하여 예외 처리를 할 수 있음

2. try문

   가) 오류가 발생할 가능성이 있는 코드 실행

   나) 예외가 발생되지 않으면 except 없이 실행 종료

3. except문

   가) 예외가 발생하면 except절이 실행

   나) 예외 상황을 처리하는 코드를 받아서 적절한 조치를 취함

4. 처리 순서

   가) 예외없는 정상종료 :  try - else - Finally

   나) 예외처리 할 경우 : try - except - Finally

   다) 예외처리 하지 못한 경우 : try - Finally  오류메시지 출력

5. 작성 방법

   ```python
   try: # try문은 반드시 한 개 이상의 except문이 필요 # 코드 실행
       try 명령문
   except 예외그룹-1 as 변수-1: # try문에서 예외 발생 시 실행
       예외처리 명령문 1
   except 예외그룹-2 as 변수-2:
       예외처리 명령문 2
   else: # try문에서 예외가 발생하지 않으면 실행
       else 명령문
   finally: # 선택사항 # 예외 발생 여부와 관계없이 항상 실행
       finally 명령문
   ```

   ```python
   '''
   가. 파일을 열고 읽는 코드를 작성
    1. 파일 열기 시도
     가) 파일 없는 경우 => '해당 파일이 없습니다.' 출력(except)
     나) 파일 있는 경우 => 파일 내용을 출력(else)
    2. 해당 파일 읽기 작업 종료 메시지 출력(finally)
   '''
   
   # 파일이 없는 경우
   try:
       f = open('nooofile.txt')
   except FileNotFoundError:
       print('해당 파일이 없습니다.')
   else:
       print('파일을 읽기 시작합니다.')
       print(f.read())
       print('파일을 모두 읽었습니다.')
       f.close()
   finally:
       print('파일 읽기를 종료합니다.')
   '''
   해당 파일이 없습니다.
   파일 읽기를 종료합니다.
   '''
   
   # 파일이 존재하는 경우
   try:
       f = open('nooofile.txt')
   except FileNotFoundError:
       print('해당 파일이 없습니다.')
   else:
       print('파일을 읽기 시작합니다.')
       print(f.read())
       print('파일을 모두 읽었습니다.')
       f.close()
   finally:
       print('파일 읽기를 종료합니다.')
   '''
   파일을 읽기 시작합니다.
   파일내용
   파일을 모두 읽었습니다.
   파일 읽기를 종료합니다.
   '''
   ```

6. 예외 처리 예시

   ```python
   try:
       num = input('숫자입력 :')
       print(int(num))
   except ValueError: # except: 라고만 써도 됨
       print('숫자가 입력되지 않았습니다.')
   ```

7. 에러 메시지 처리(as) : as 키워드를 활용하여 원본 에러 메시지를 사용할 수 있음, 예외를 다른 이름에 대입

   ```python
   [][1] # IndexError : list index out of range
   
   try:
       empty_list = []
       print(empty_list[-1])
   except IndexError as err:
       print(f'{err}, 오류가 발생했습니다.') # list index out of range, 오류가 발생했습니다.
   ```

8. 복수의 예외 처리

   ```python
   # 100을 사용자가 입력한 값으로 나누고 출력하는 코드 작성
   
   # 합하여 처리
   try:
       num = input('100으로 나눌 값을 입력하시오 : ')
   	print(100/int(num))
   except (ValueError, ZeroDivisonError): # 100/int('a') -> 문자열을 int로 형변환
       print('제대로 입력해줘.')
       
   # 별도 처리, 순차적으로 수행됨으로 가장 작은 범주부터 예외 처리 해야함
   try:
       num = input('100으로 나눌 값을 입력하시오 : ')
   	print(100/int(num))
   except ValueError:
       print('숫자를 넣어주세요.')
   except ZeroDivisonError:
       print('0으로 나눌 수 없습니다.')
   except:
       print('에러는 모르지만 에러가 발생하였습니다.')
   ```

   